pyqtorch/analog.py

```diff
@@ -57,62 +79,132 @@ def __init__(self, operations: list[Module]):
         return reduce(add, (op(state, values) for op in self.operations))
 
+    def tensor(self, values: dict = {}, n_qubits: int = 1) -> Tensor:
+        mat = torch.zeros((2, 2, 1), device=self.device)
+        for _ in range(n_qubits - 1):
+            mat = torch.kron(mat, torch.zeros((2, 2, 1), device=self.device))
+        return reduce(add, (op.tensor(values, n_qubits) for op in self.operations), mat)
 
-class Hamiltonian(Add):
-    def __init__(self, operations: list[torch.nn.Module]):
-        if all([not isinstance(op, (Parametric)) for op in operations]):
-            super().__init__(operations)
-        else:
-            raise TypeError(
-                "Hamiltonian can only contain the following operations: [Primitive, Scale, Add]."
-            )
+
+TGenerator = Union[torch.nn.ModuleList, list, Tensor, Primitive]
 
 
-class HamiltonianEvolution(torch.nn.Module):
+class Hamiltonian(Add):
     def __init__(
         self,
+        generator: TGenerator,
         qubit_support: Tuple[int, ...],
     ):
-        super().__init__()
-        self.qubit_support: Tuple[int, ...] = qubit_support
+        assert isinstance(
+            generator, (Tensor, Primitive, Sequence)
+        ), "Generator can be: primitive, tensor or sequence"
+        if isinstance(generator, Tensor):
+            generator = [Primitive(generator, target=qubit_support[0])]
+        elif isinstance(generator, Primitive):
+            generator = [generator]
+        super().__init__(operations=generator)
+        self.qubit_support = qubit_support
+        self.generator = self.operations
+
+
+def is_diag_hamiltonian(hamiltonian: Tensor) -> bool:
+    diag_check = torch.tensor(
+        [is_diag(hamiltonian[..., i]) for i in range(hamiltonian.shape[BATCH_DIM])],
+        device=hamiltonian.device,
+    )
+    return bool(torch.prod(diag_check))
+
+
+def evolve(hamiltonian: Operator, time_evolution: torch.Tensor) -> Tensor:
+    if is_diag_hamiltonian(hamiltonian):
+        evol_operator = torch.diagonal(hamiltonian) * (-1j * time_evolution).view((-1, 1))
+        evol_operator = torch.diag_embed(torch.exp(evol_operator))
+    else:
+        evol_operator = torch.transpose(hamiltonian, 0, -1) * (-1j * time_evolution).view(
+            (-1, 1, 1)
+        )
+        evol_operator = torch.linalg.matrix_exp(evol_operator)
+    return torch.transpose(evol_operator, 0, -1)
 
-        def _diag_operator(hamiltonian: Operator, time_evolution: torch.Tensor) -> Operator:
-            evol_operator = torch.diagonal(hamiltonian) * (-1j * time_evolution).view((-1, 1))
-            evol_operator = torch.diag_embed(torch.exp(evol_operator))
-            return torch.transpose(evol_operator, 0, -1)
 
-        def _matrixexp_operator(hamiltonian: Operator, time_evolution: torch.Tensor) -> Operator:
-            evol_operator = torch.transpose(hamiltonian, 0, -1) * (-1j * time_evolution).view(
-                (-1, 1, 1)
+class HamiltonianEvolution(Sequence):
```

Comment:
@Roland-djee @jpmoutinho my proposal for the new HamEvo which will handle everything we need in qadence. not finished yet. mainly block_to_tensor-related things have to be resolved still